src/value_generators/src/DocumentGenerator.cpp

```diff
@@ -1173,6 +1190,154 @@ ChooseGenerator::ChooseGenerator(const Node& node, GeneratorArgs generatorArgs)
+ * @return
+ *   a constant generator if given a constant/scalar.
+ */
+UniqueGenerator<int64_t> doubleTimeGenerator(const Node& node, GeneratorArgs generatorArgs) {
+    auto generator = doubleGenerator(node, generatorArgs);
+    return std::make_unique<DoubleToIntGenerator>(generator);
+}
+
+/**
+ * @private
+ * @param node
+ *   a node containing a ^Now or ^RandomDate.
+ * @return
+ *   a constant generator if given a constant/scalar.
+ * @throws
+ *   InvalidConfigurationException if the node doesn't hold a ^Now or ^RandomDate.
+ */
+UniqueGenerator<int64_t> dateTimeGenerator(const Node& node, GeneratorArgs generatorArgs) {
+    // Set of parsers to look when we request a date parser
+    const static std::map<std::string, Parser<UniqueGenerator<bsoncxx::types::b_date>>> dateParsers{
+        {"^Now",
+         [](const Node& node, GeneratorArgs generatorArgs) {
+             return std::make_unique<NowGenerator>(node, generatorArgs);
+         }},
+        {"^RandomDate",
+         [](const Node& node, GeneratorArgs generatorArgs) {
+             return std::make_unique<RandomDateGenerator>(node, generatorArgs);
+         }},
+    };
+
+    if (auto parserPair = extractKnownParser(node, generatorArgs, dateParsers)) {
+        // known parser type
+        auto generator = parserPair->first(node[parserPair->second], generatorArgs);
+        return std::make_unique<DateToIntGenerator>(generator);
+    }
+    // This is an internal private function, we have no sane way to interpret a scalar here,
+    // so throw an exception.
+    std::stringstream msg;
+    msg << "Unknown parser: not an expected type " << node;
+    BOOST_THROW_EXCEPTION(InvalidValueGeneratorSyntax(msg.str()));
+}
+
+/**
+ * @param node
+ *   a node containing a date generator.
+ * generator
+ * @return
+ *   a constant generator if given a constant/scalar.
+ */
+UniqueGenerator<int64_t> dateGenerator(const Node& node,
+                                       GeneratorArgs generatorArgs,
+                                       const boost::posix_time::ptime& default_time) {
+    const static auto timeGenerators = {
+        dateTimeGenerator, dateStringTimeGenerator, intGenerator, doubleTimeGenerator};
+    if (node) {
+        for (auto&& generator : timeGenerators) {
+            try {
+                return generator(node, generatorArgs);
+            } catch (const InvalidValueGeneratorSyntax& e) {
+                if (!isUnknownParser(e) && !isUnknownDateFormat(e)) {
+                    throw e;
+                }
+            } catch (const InvalidConversionException& e) {
+            }
+        }
+        std::stringstream msg;
+        msg << "Unknown parser: not an expected type " << node;
+        BOOST_THROW_EXCEPTION(InvalidValueGeneratorSyntax(msg.str()));
+    }
+
+    // No value, get the appropriate default.
+    auto millis = (default_time - epoch).total_milliseconds();
+    return std::make_unique<ConstantAppender<int64_t>>(millis);
```

Comment:
Does a lot of this of this go away if we don't allow mixed types in the min/max fields? I.e. if you want a literal min/max rather than generated simply require the use of `^Verbatim`? What you have here is clever and well-written, but I wonder if it's complexity worth having.